text/0031-transactions-with-contingent-inputs.md

```diff
@@ -0,0 +1,191 @@
+to permit swapping bitcoin in a "trustless" way for other assets, using hash time locked contracts. These protocols are often quite complex,
+requiring each party to "commit" their funds first, then trigger the swap, and have contingencies for each party to recover their committed funds
+if the counter party renegs. These protocols may require several rounds of interaction, and require several on-chain transactions, incurring fees and
+increasing the total delay for the swap to be finished.
+
+This proposal is considerably simpler -- the rounds of interaction are a minimum, requiring only one side to send a transaction fragment to another.
+No one is required to "escrow" their funds into a smart contract or temporary address. A party who desires to swap can broadcast their transaction fragment
+to many potential counterparties. Because the Ring MLSAG contains the key image of the spent input, everyone is able to check if the transaction fragment is
+still valid, and the initiating party can cancel the swap proposal by simply spending input to themselves in a separate transaction, burning its key image and preventing any of the
+broadcasted transaction fragments from being used in a valid transaction.
+
+A feature like this could potentially be used to build a fast, clean asset-swap experience in a mobile app, without relying on smart contracting functionality.
+
+More generally, it may be interesting to explore ways that users can collaboratively sign transactions securely. This is one way that we can experiment with this,
+and could potentially be extended in the future.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In a MobileCoin transaction, an input consists of a collection of `TxOut`'s in the ledger, one of which is truly being spent and others of which are just being mixed in. This ring,
+and the proofs of membership, appear in the `TxIn`, in the `TxPrefix`. A transaction produces several new `TxOut`'s, and these outputs appear in the `TxPrefix`.
+
+The signature part of a transaction contains several things:
+* A `RingMLSAG` for each `TxIn`
+* A `pseudo_output_commitment` for each `TxIn`
+* A range proof for each `pseudo_output_commitment`, relative to the generator for this token id.
+
+Transaction validation ensures that the sum of outputs minus the sum of pseudo-outputs is the zero point on the elliptic curve, (including the implicit fee output)
+which ensures that the transaction is valid.
+
+This feature contains two discrete steps:
+* Allowing mixed transactions. This requires that individual outputs and pseudo-outputs in a transaction can be labeled with token ids.
+  (The MCIP #25 token id, added to the tx prefix, is now interpretted as the fee token id.)
+  When validating range proofs, each individual output and pseudo-output has its range proof checked relative to its specified token id.
+  No other changes are required to support mixed transactions.
```

Comment:
And the privacy properties of the on-chain TxOuts also don't change, correct? As this is all in the TxPrefix, which, as discussed in MCIP #25, is delivered through an attested channel and is only visible by the enclave?

Reply:
Yup!